### PR TITLE
Review upload

### PR DESCRIPTION
**Pull request by vitalik12345678**
dev
## JIRA

* [Main JIRA ticket](https://jira.softserve.academy/secure/RapidBoard.jspa?rapidView=id)

## Code reviewers

- [ ] @github_username

### Second Level Review

- [ ] @github_username

## Summary of issue

ToDo

## Summary of change

ToDo

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

**Review by VictoriaAmazonka — CHANGES_REQUESTED** [review]
HTTP GET doesn’t have body

File: src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java [hunk 1]
```diff
@@ -103,25 +120,135 @@ public void deleteImages(String urlLogo, String urlBackground, List<GalleryPhoto
...
+        } catch (IOException ioe) {
+            try {
+                FileUtils.forceDelete(file);
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+            throw new FileUploadException(String.format(FILE_UPLOAD_EXCEPTION, fileName));
+        }
+
+        if (file.exists()){
```
Comment (VictoriaAmazonka) [inline src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java:225]: Whitespace

**Review by VictoriaAmazonka — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java [hunk 2]
```diff
@@ -103,25 +120,135 @@ public void deleteImages(String urlLogo, String urlBackground, List<GalleryPhoto
...
+        File file = new File(TEMP_FILE_STORAGE+fileUpdateProfile.getFileName());
+        saveFile(uploadDir, fileUpdateProfile.getFileName(), file,fileUpdateProfile.getBase64());
+
+        uploadDir = uploadDir+"/"+ fileUpdateProfile.getFileName();
+        uploadDir = uploadDir.substring(uploadDir.indexOf("/"));
+
+        galleryPhoto.setUrl(uploadDir);
+        galleryRepository.save(galleryPhoto);
+
+        return "File changed successful";
```
Comment (VictoriaAmazonka) [inline src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java:157]: It’s not good idea to return literal String. Please, return boolean

**Review by VictoriaAmazonka — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java [hunk 3]
```diff
@@ -103,24 +114,132 @@ public void deleteImages(String urlLogo, String urlBackground, List<GalleryPhoto
...
+    }
+
+    private void saveFile(String uploadDir, String fileName, File file, String byte64) {
+        Path uploadPath = Paths.get(uploadDir);
+
+        InputStream inputStream = null;
+        try {
+            inputStream = new FileInputStream(file);
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
```
Comment (VictoriaAmazonka) [inline src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java:161]: Logger

File: src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java [hunk 4]
```diff
@@ -103,24 +114,132 @@ public void deleteImages(String urlLogo, String urlBackground, List<GalleryPhoto
...
+        }
+
+        if (file.length() > IMAGE_SIZE_B) {
+            try {
+                file.deleteOnExit();
+                FileUtils.forceDelete(file);
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (VictoriaAmazonka) [inline src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java:169]: Logger

File: src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java [hunk 5]
```diff
@@ -103,24 +114,132 @@ public void deleteImages(String urlLogo, String urlBackground, List<GalleryPhoto
...
+
+        try {
+            if (!Files.exists(uploadPath)) {
+                Files.createDirectories(uploadPath);
+            }
+        } catch (IOException e) {
+            try {
+                FileUtils.forceDelete(file);
+            } catch (IOException ioException) {
+                ioException.printStackTrace();
```
Comment (VictoriaAmazonka) [inline src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java:202]: Logger

File: src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java [hunk 6]
```diff
@@ -146,4 +265,36 @@ private int ordinalIndexOf(final String str, final String searchStr, final int o
...
+        FileOutputStream fileOutputStream;
+        try {
+            fileOutputStream = new FileOutputStream("./src/tempImage/" + fileName);
+            fileOutputStream.write(decod);
+            fileOutputStream.close();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+
+        Path path = Paths.get("src/tempImage/");
```
Comment (VictoriaAmazonka) [inline src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java:291]: Why do you save to folder src? This folder stays for source code

**Review by VictoriaAmazonka — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java [hunk 7]
```diff
@@ -146,4 +265,36 @@ private int ordinalIndexOf(final String str, final String searchStr, final int o
...
+        if (!Files.exists(path)) {
+            try {
+                Files.createDirectories(path);
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (VictoriaAmazonka) [inline src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java:296]: Logger

File: src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java [hunk 8]
```diff
@@ -146,4 +265,36 @@ private int ordinalIndexOf(final String str, final String searchStr, final int o
...
+        Base64.Decoder decoder = Base64.getMimeDecoder();
+        byte[] decod = decoder.decode(base64);
+
+        FileOutputStream fileOutputStream;
+        try {
+            fileOutputStream = new FileOutputStream("./src/tempImage/" + fileName);
+            fileOutputStream.write(decod);
+            fileOutputStream.close();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (VictoriaAmazonka) [inline src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java:288]: Logger

**Review by VictoriaAmazonka — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java (hunk 6 above)
Comment (VictoriaAmazonka) [inline src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java:291]: Literals, create constants

**Review by VictoriaAmazonka — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java [hunk 9]
```diff
@@ -146,4 +286,49 @@ private int ordinalIndexOf(final String str, final String searchStr, final int o
...
+        encode(uploadProfile.getBase64(), uploadProfile.getFileName());
+    }
+
+    private void encode(String base64, String fileName) {
+        Path folderPath = Paths.get(TEMP_FILE_STORAGE);
+        if (!Files.exists(folderPath)) {
+            try {
+                Files.createDirectories(folderPath);
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (VictoriaAmazonka) [inline src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java:303]: Logger

**Review by VictoriaAmazonka — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java [hunk 10]
```diff
@@ -1,85 +1,104 @@
...
+
+        File file = new File(TEMP_FILE_STORAGE + uploadProfile.getFileName());
+        String fileName = StringUtils.cleanPath(file.getName());
+        saveFile(uploadDir, fileName, file, uploadProfile.getBase64());
 
         String actualPath = String.format("/%s/%s", uploadDir, fileName);
+        Optional<Club> optionalClub = clubRepository.findById(uploadProfile.getId());
+        Club club = optionalClub.get();
+        GalleryPhoto galleryPhoto = new GalleryPhoto();
+        actualPath = actualPath.substring(7);
```
Comment (VictoriaAmazonka) [inline src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java:90]: What is 7? Some magic number

**Review by VictoriaAmazonka — CHANGES_REQUESTED** [review]
Please, review all comments and do changes

File: src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java [hunk 11]
```diff
@@ -60,40 +58,49 @@ public FileUploadServiceImpl(GalleryRepository galleryRepository, ClubRepository
...
     @Override
     public String getPhoto(String filePath) {
-        GalleryPhoto galleryPhoto = galleryRepository.findByUrl(filePath);
-        File file = new File(TARGET + filePath);
-        try {
-            byte[] baseImage = FileUtils.readFileToByteArray(file);
-            return Base64.getEncoder().encodeToString(baseImage);
-        } catch (IOException e) {
-            log.error(e.getMessage(),e);
+        Optional<GalleryPhoto> galleryPhoto = Optional.ofNullable(galleryRepository.findByUrl(filePath));
```
Comment (RomanKlymus) [inline src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java:61]: Move to repository creating of the Optional object to the method. findByUrl must return Optional<String>.

File: src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java [hunk 12]
```diff
@@ -60,40 +58,49 @@ public FileUploadServiceImpl(GalleryRepository galleryRepository, ClubRepository
...
+            try {
+                byte[] baseImage = FileUtils.readFileToByteArray(file);
+                Optional<String> s = Optional.ofNullable(Base64.getEncoder().encodeToString(baseImage));
+                return s.isPresent() ? s.get() : "null";
+            } catch (IOException e) {
+                log.error(e.getMessage(), e);
+                log.debug("get file from path" + filePath);
+                throw new FileUploadException("Some problems while file reading ");
+            }
+        }else{
```
Comment (RomanKlymus) [inline src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java:73]: Check-in if statement !galleryPhoto.isPresent() and then throw exception. All else code write outside of the body of if statement.

File: src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java [hunk 13]
```diff
@@ -60,40 +58,49 @@ public FileUploadServiceImpl(GalleryRepository galleryRepository, ClubRepository
...
-            return Base64.getEncoder().encodeToString(baseImage);
-        } catch (IOException e) {
-            log.error(e.getMessage(),e);
+        Optional<GalleryPhoto> galleryPhoto = Optional.ofNullable(galleryRepository.findByUrl(filePath));
+        if (galleryPhoto.isPresent()) {
+            File file = new File(TARGET + filePath);
+            try {
+                byte[] baseImage = FileUtils.readFileToByteArray(file);
+                Optional<String> s = Optional.ofNullable(Base64.getEncoder().encodeToString(baseImage));
+                return s.isPresent() ? s.get() : "null";
```
Comment (RomanKlymus) [inline src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java:67]: Why do you check here is an object not null. And don't return null as a string!!!!!!!!!
It is useless to check if an object equals and then return null. You must do something else. Throw exception for example.

File: src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java [hunk 14]
```diff
@@ -60,40 +58,49 @@ public FileUploadServiceImpl(GalleryRepository galleryRepository, ClubRepository
...
-
-        String uploadDir = String.format("%s/%s", uploadDirectory, uploadProfile.getFolder());
-        encodeBase64(uploadProfile);
-
-        File file = new File(TEMP_FILE_STORAGE + uploadProfile.getFileName());
-        String fileName = StringUtils.cleanPath(file.getName());
-        saveFile(uploadDir, fileName, file, uploadProfile.getBase64());
-
-        String actualPath = String.format("/%s/%s", uploadDir, fileName);
         Optional<Club> optionalClub = clubRepository.findById(uploadProfile.getId());
```
Comment (RomanKlymus) [inline src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java:81]: Same as above, do optional in repo

**Comment by vitalik12345678** [comment]
Cleaned code in file upload service,added some check in methods

File: src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java [hunk 15]
```diff
@@ -58,49 +59,32 @@ public FileUploadServiceImpl(GalleryRepository galleryRepository, ClubRepository
...
-        if (galleryPhoto.isPresent()) {
             File file = new File(TARGET + filePath);
-            try {
-                byte[] baseImage = FileUtils.readFileToByteArray(file);
-                Optional<String> s = Optional.ofNullable(Base64.getEncoder().encodeToString(baseImage));
-                return s.isPresent() ? s.get() : "null";
+        try {
+            byte[] baseImage = FileUtils.readFileToByteArray(file);
+            Optional<String> s = Optional.ofNullable(Base64.getEncoder().encodeToString(baseImage));
+            return s.isPresent() ? s.get() : s.orElseThrow(IOException::new);
```
Comment (RomanKlymus) [inline src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java:66]: change this whole line to return s.orElseThrow(IOException::new);

Comment (RomanKlymus) [reply src/main/java/com/softserve/teachua/service/impl/FileUploadServiceImpl.java:66]: And why do you throw IOException? Throw exception which will be handled by the handler and return 404 HTTP status.

**Pull request closed without merging**